File: 01 solved.ac (baekjoon)/02. Silver/q9095.java
import java.io.*;

public class q9095 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        br.readLine();

        StringBuilder sb = new StringBuilder();
        while(br.ready()) {
            int num = Integer.parseInt(br.readLine());
            int total = backTrack(num, 0);
            sb.append(total).append("\n");
        }

        System.out.print(sb);
    }

    private static int backTrack(int target, int current) {

        // 종료 조건 + 현재 동작
        // 전처리 + 다음 동작(재귀)

        if (current == target) return 1;
        if (current > target) return 0;

        int count = 0;
        for (int i = 1; i <= 3; i++) {  // 1, 2, 3
            count += backTrack(target, current + i);
        }

        return count;
    }
}

File: 01 solved.ac (baekjoon)/02. Silver/q9095.py
"""백준 9095. 1, 2, 3 더하기"""

def backTrack(target, current) :
    """back tracking for target"""
    if current == target : return 1
    if current > target : return 0

    temp_count = 0
    for i in range(1, 4) :
        temp_count += backTrack(target, current + i)

    return temp_count


numbers = int(input())

for _ in range(numbers) :
    num = int(input())
    total = backTrack(num, 0)
    print(total)
